### ISSUE TITLE
Issue with order of operations with DISTINCT

### ISSUE THREAD

**Issue by iangow** [tidyverse/dbplyr#1179]
## Brief description of the problem

I expect the same answer from all three queries below. The queries are the same except for the following

1. Using local data frame
2. Using SQL
3. Using SQL with `compute()` to force evaluation of `distinct()`

However, the second one gives a different answer. This is a new-ish issue, as code that worked previously is now broken.

I think the translated SQL should be something like

```sql
SELECT `name`, `id`,
    COUNT(*) OVER (PARTITION BY `name`) AS `n_distinct`
FROM (SELECT DISTINCT `name`, `id` FROM `df`) AS some_table
```

## Reprex

``` r
library(tidyverse)

set.seed(2023)

df <- 
  tibble(name = letters, id = 1:26) %>%
  mutate(rep = sample(list(1, 1:2, 1:3), 
                       size = nrow(.), replace = TRUE)) %>%
  unnest(rep) 

df
#> # A tibble: 45 × 3
#>    name     id   rep
#>    <chr> <int> <dbl>
#>  1 a         1     1
#>  2 b         2     1
#>  3 b         2     2
#>  4 b         2     3
#>  5 c         3     1
#>  6 d         4     1
#>  7 d         4     2
#>  8 d         4     3
#>  9 e         5     1
#> 10 e         5     2
#> # … with 35 more rows

# Query 1: dplyr
df %>%
  select(name, id) %>%
  distinct() %>%
  group_by(name) %>%
  mutate(n_distinct = n()) %>%
  ungroup() %>%
  filter(n_distinct > 1)
#> # A tibble: 0 × 3
#> # … with 3 variables: name <chr>, id <int>, n_distinct <int>

con <- DBI::dbConnect(RSQLite::SQLite(), dbname = ":memory:")

df_sql <- copy_to(con, df)

# Query 2: SQL
df_sql %>%
  select(name, id) %>%
  distinct() %>%
  group_by(name) %>%
  mutate(n_distinct = n()) %>%
  ungroup() %>%
  filter(n_distinct > 1)
#> # Source:   SQL [?? x 3]
#> # Database: sqlite 3.40.1 [:memory:]
#>    name     id n_distinct
#>    <chr> <int>      <int>
#>  1 b         2          3
#>  2 d         4          3
#>  3 e         5          2
#>  4 f         6          2
#>  5 n        14          3
#>  6 o        15          2
#>  7 p        16          3
#>  8 q        17          3
#>  9 u        21          3
#> 10 w        23          2
#> # … with more rows

# SQL for problematic Query 2
df_sql %>%
  select(name, id) %>%
  distinct() %>%
  group_by(name) %>%
  mutate(n_distinct = n()) %>%
  ungroup() %>%
  show_query()
#> <SQL>
#> SELECT DISTINCT
#>   `name`,
#>   `id`,
#>   COUNT(*) OVER (PARTITION BY `name`) AS `n_distinct`
#> FROM `df`

# Query 3: SQL with compute to force evaluation of distinct()
df_sql %>%
  select(name, id) %>%
  distinct() %>%
  compute() %>%
  group_by(name) %>%
  mutate(n_distinct = n()) %>%
  ungroup() %>%
  filter(n_distinct > 1)
#> # Source:   SQL [0 x 3]
#> # Database: sqlite 3.40.1 [:memory:]
#> # … with 3 variables: name <chr>, id <int>, n_distinct <lgl>
```

<sup>Created on 2023-03-04 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by mgirlich** [comment]
Thanks for reporting the issue. It is already fixed in dbplyr 2.3.1

**Comment by iangow** [comment]
Thanks. I looked for a similar issue, but somehow didn't find #1119.
